csrc/preseg_passes/allocation_order_inference.cpp

```diff
@@ -15,442 +16,279 @@ namespace nvfuser::preseg_passes {
-  //    |
-  //    Operation
-  //    |
-  //    v
-  //   consumer rfactor dom [..., i0, ..., i1, ..., i2, ...]
-  //
-  // we construct allocation domain on producer, filtering out reduction, apply
-  // root domain map from producer to consumer.
-  //   [r0', i0', i2', i1'] -> [i0', i2', i1'] -> [i0, i2, i1]
-  // so the function would return [i0, i2, i1]
-  std::vector<IterDomain*> propagateAllocationDomain(
-      TensorView* producer,
-      TensorView* consumer) {
-    // constructing alloc_domain for producer from its root domain, while
-    // filtering out reduction because they won't appear in consumer's domain.
-    std::vector<IterDomain*> alloc_domain = TensorDomain::noReductions(
-        constructAllocationDomain(producer, alloc_order_map_.at(producer)));
-    // creating producer to consumer root domain map
-    std::unordered_map<IterDomain*, IterDomain*> p2c_map =
-        PairwiseRootDomainMap(producer, consumer).mapProducerToConsumer();
-    // map alloc_domain to consumer
-    std::transform(
-        alloc_domain.cbegin(),
-        alloc_domain.cend(),
-        alloc_domain.begin(),
-        [&p2c_map](IterDomain* id) { return p2c_map.at(id); });
-    return alloc_domain;
-  }
-
-  // Propagate allocation order from producer to consumer via:
-  // 1. Constructs producer allocation_domain with its allocation order;
-  // 2. Mapping it to consumer's root domain to create alloc_domain;
-  // 3. Compute allocation order of consumer as the permutation between
-  //    alloc_domain and `permutation_ref`.
-  //
-  // Returns true when producer has a recorded allocation order, false
-  // otherwise. This function assumes that all root domain in consumer can be
-  // mapped to producer.
-  bool propagateAllocationOrder(
-      TensorView* producer,
-      TensorView* consumer,
-      const std::vector<IterDomain*>& permutation_ref) {
-    auto iter = alloc_order_map_.find(producer);
-    // early return is producer doesn't have an entry in alloc_order_map_
-    if (iter == alloc_order_map_.end()) {
-      return false;
-    }
-
-    // early termination to propagate empty allocation order
-    if (iter->second.empty()) {
-      alloc_order_map_[consumer] = {};
-      return true;
+// Note: we could be using a simplified logic below,
+// See issue https://github.com/NVIDIA/Fuser/issues/2202
+// 1. we project iter domains from targets' rfactor domain which has an exact
+// map to ref's allocation domain.
+//   mapped_id_vec {ir5[i1], iS7[i2]}
+// 2. remove all projected iter domains from target's rfactor
+// domain:
+//   unmapped_ids_vec {iS3[i3], iS4[i4], iS6[i5], ir8[1]}
+// 3. append mapped_id_vec at the end of unmapped_id_vec.
+//   target_alloc_domain
+//   {iS3[i3], iS4[i4], iS6[i5], ir8[1], ir5[i1], iS7[i2]}
+void AllocationOrderMapping(
+    const IdModel& id_model,
+    const TensorView* ref,
+    TensorView* target) {
+  const DisjointSets<Val*>& val_sets =
+      id_model.idGraph(IdMappingMode::EXACT).disjointValSets();
```

Comment:
My motivation is that we do NOT want to map broadcast IDs. I think we should avoid mapping trivial iterdomain. (be it broadcast or reduction).

A concrete (likely limited) example:
If we are trying to map TV0 [i0, i1, b2, i3] to TV1 [i0, i1, i4, i3]. where broadcast `b2` would map to `i4`

if TV0 has an allocation domain as {i0, i1, b2, i3}, if we do NOT map broadcast iterdomain, then the target TV1 would have an allocation domain of {i4, i0, i1, i3}, which has a higher upper bound for vectorization for TV0 {i0*i1*i3} (talking about a point-wise scheduler).
otherwise if we map broadcast, the allocation order would be {i0, i1, i4, i3}. TV0 would only have {i3} as the upper bound for vectorization.

Reply:
Hmm, if you are talking about pointwise fusion, wouldn't TV0 be vectorizable to `i3` already? For example, if we have
```
T0[1024, 1024, 1, 3] + T1[1024, 1024, 4, 3]
```
We can not vectorize 4 for T0 due to loop promotion.

Reply:
I'm assuming ` loop promotion` means inlining?!

> T0[1024, 1024, 1, 3] + T1[1024, 1024, 4, 3]

This is a bad example. :) Because T1 has more LoopIterDomain so it would dominate the output stride order.

I'm going to change it to > T2[1025, 1024, 4, 3] = T0[1025, 1024, 1, 3] + T1[1, 1, 4, 1]

if we propagate T0's allocation domain to T2 while ignoring broadcast, We'll have T2's allocation domain being
```T2_alloc = [4, 1025, 1024, 3]```

So scheduling will reorder with `T2[4, 1025, 1024, 3]` which will propagate to `T0[1, 1025, 1024, 3]`. giving it's allocation domain `T0_alloc[1025, 1024, 1, 3]`, shouldn't we be able to figure out that we [1025, 1024, 3] is contiguous and we should be able to vectorize on the factor of [1025*1024*3]?

Reply:
Oh, I see what you mean. You are right. Thanks for the clarification!